src/main/scala/edu/alexy/lectures/concurrency/RoboAdviser.scala

```diff
@@ -2,50 +2,95 @@ package edu.alexy.lectures.concurrency
 import edu.alexy.lectures.concurrency.model.Company
 import edu.alexy.lectures.concurrency.util.{Db, DbError, WebApi, WebApiError}
-import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
 
+import scala.util.{Failure, Success}
+import scala.concurrent.ExecutionContext.Implicits.global
+
 object RoboAdviser {
   // Task 1.
   // Return 'AAPL' revenue from `Db.getCompanyLastFinancials`. Possible error should be returned as a ServiceError.
-  def getAAPLRevenue: Future[Double] = ???
+  def getAAPLRevenue: Future[Double] = Db.getCompanyLastFinancials("AAPL").transform {
+    case Success(value) => Success(value.map(_.revenue.toDouble).getOrElse(0.0))
+    case Failure(_) => Failure(DbError)
+  }
 
   // Task 2.
   // Implement a fallback strategy for 'Db.getAllTickers'.
   // 'Db.getAllTickers' should be called no more then 'retries' times.
   // Possible error should be returned as a ServiceError.
-  def getAllTickersRetryable(retries: Int = 10): Future[Seq[String]] = ???
+  def getAllTickersRetryable(retries: Int = 10): Future[Seq[String]] = {
+    val res = Db.getAllTickers
+    if (retries <= 0) res.transform {
+      case Success(value) => Success(value)
+      case Failure(_) => Failure(DbError)
+    } else res.fallbackTo(getAllTickersRetryable(retries - 1))
+  }
 
   // Task 3.
   // Implement a fallback strategy for 'Db.getCompanyLastFinancials'.
   // 'Db.getCompanyLastFinancials' should be called no more then 'retries' times.
   // Possible error should be returned as a ServiceError.
-  def getCompanyRetryable(ticker: String, retries: Int = 10): Future[Option[Company]] = ???
+  def getCompanyRetryable(ticker: String, retries: Int = 10): Future[Option[Company]] = {
+    val res = Db.getCompanyLastFinancials(ticker)
+    if (retries <= 0) res.transform {
+      case Success(value) => Success(value)
+      case Failure(_) => Failure(DbError)
+    } else res.fallbackTo(getCompanyRetryable(ticker, retries - 1))
+  }
 
   // Task 4.
   // Implement a fallback strategy 'WebApi.getPrice'.
   // 'WebApi.getPrice' should be called no more then 'retries' times.
   // Possible error should be returned as a ServiceError.
-  def getPriceRetryable(ticker: String, retries: Int = 10): Future[Double] = ???
+  def getPriceRetryable(ticker: String, retries: Int = 10): Future[Double] = {
+    val res: Future[Double] = WebApi.getPrice(ticker)
+    if (retries <= 0) res.transform {
+      case Success(value) => Success(value)
+      case Failure(_) => Failure(WebApiError)
+    } else res.fallbackTo(getPriceRetryable(ticker, retries - 1))
+  }
 
   // Task 5.
   // Using retryable functions return all tickers with their real time prices.
-  def getAllTickersPrices: Future[Seq[(String, Double)]] = ???
+  def getAllTickersPrices: Future[Seq[(String, Double)]] = {
+    val tickers: Future[Seq[String]] = getAllTickersRetryable()
+    val prices: Future[Seq[Double]] = tickers.flatMap(x => Future.sequence(x.map(getPriceRetryable(_))))
```

Comment:
You can return a tuple of ticker and price by mapping against `getPriceRetryable`. for-comp becomes redundant afterward.